doc: add path rules and validation for export targets in package.json (#58604)
nodejs/node: doc/api/packages.md

in relative and absolute import specifiers.

#### Path Rules and Validation for Export Targets

When defining paths as targets in the [`"exports"`][] field, Node.js enforces
several rules to ensure security, predictability, and proper encapsulation.
Understanding these rules is crucial for authors publishing packages.

##### Targets must be relative URLs

All target paths in the [`"exports"`][] map (the values associated with export
keys) must be relative URL strings starting with `./`.

```json
// package.json
{
  "name": "my-package",
  "exports": {
    ".": "./dist/main.js",          // Correct
    "./feature": "./lib/feature.js", // Correct
    // "./origin-relative": "/dist/main.js", // Incorrect: Must start with ./
    // "./absolute": "file:///dev/null", // Incorrect: Must start with ./
    // "./outside": "../common/util.js" // Incorrect: Must start with ./
  }
}
```

Reasons for this behavior include:

* **Security:** Prevents exporting arbitrary files from outside the
  package's own directory.
* **Encapsulation:** Ensures all exported paths are resolved relative to
  the package root, making the package self-contained.

> #### No path traversal or invalid segments

Reviewer: ```suggestion
##### No path traversal or invalid segments
```

export. This also mirrors the requirement of using [the full specifier path][]
in relative and absolute import specifiers.

#### Path Rules and Validation for Export Targets

When defining paths as targets in the [`"exports"`][] field, Node.js enforces
several rules to ensure security, predictability, and proper encapsulation.
Understanding these rules is crucial for authors publishing packages.

**1. Targets Must Be Relative and Start with `./`**

All target paths in the [`"exports"`][] map (the values associated with export

> keys) **must** be relative path strings starting with `./`.

Reviewer: ```suggestion
keys) must be relative URL strings starting with `./`.
```

```json
// package.json
{
  "name": "my-package",
  "exports": {
    ".": "./dist/main.js",          // Correct
    "./feature": "./lib/feature.js", // Correct

>     // "absolute": "/dist/main.js", // Incorrect: Must start with ./

Reviewer: ```suggestion
    // "./origin-relative": "/dist/main.js", // Incorrect: Must start with ./
    // "./absolute": "file:///dev/null", // Incorrect: Must start with ./
```

    // "outside": "../common/util.js" // Incorrect: Must start with ./
  }
}
```

> * **Why?**
>   * **Security:** Prevents exporting arbitrary files from outside the
>     package's own directory.
>   * **Encapsulation:** Ensures all exported paths are resolved relative to
>     the package root, making the package self-contained.

Reviewer: I don't think it makes sense to have a list with a single item
```suggestion
Reasons for this behavior include:

* **Security:** Prevents exporting arbitrary files from outside the
    package's own directory.
* **Encapsulation:** Ensures all exported paths are resolved relative to
    the package root, making the package self-contained.
```

> **2. No Path Traversal or Invalid Segments**

Reviewer: ```suggestion
#### No path traversal or invalid segments
```

export. This also mirrors the requirement of using [the full specifier path][]
in relative and absolute import specifiers.

#### Path Rules and Validation for Export Targets

When defining paths as targets in the [`"exports"`][] field, Node.js enforces
several rules to ensure security, predictability, and proper encapsulation.
Understanding these rules is crucial for authors publishing packages.

**1. Targets Must Be Relative and Start with `./`**

All target paths in the [`"exports"`][] map (the values associated with export
keys) **must** be relative path strings starting with `./`.

```json
// package.json
{
  "name": "my-package",
  "exports": {
    ".": "./dist/main.js",          // Correct
    "./feature": "./lib/feature.js", // Correct
    // "absolute": "/dist/main.js", // Incorrect: Must start with ./
    // "outside": "../common/util.js" // Incorrect: Must start with ./
  }
}
```

* **Why?**
  * **Security:** Prevents exporting arbitrary files from outside the
    package's own directory.
  * **Encapsulation:** Ensures all exported paths are resolved relative to
    the package root, making the package self-contained.

> **2. No Path Traversal and Invalid Segments Disallowed**

Reviewer 2: Nit:

```suggestion
**2. No Path Traversal or Invalid Segments**
```

Export targets **must not** resolve to a location outside the package's root
directory. Additionally, path segments like `.` (single dot), `..` (double dot),
or `node_modules` (and their URL-encoded equivalents) are generally disallowed
within the `target` string after the initial `./` and in any `subpath` part
substituted into a target pattern.

```json
// package.json
{
  "name": "my-package",
  "exports": {
    // ".": "./dist/../../elsewhere/file.js", // Invalid: path traversal
    // ".": "././dist/main.js",             // Invalid: contains "." segment
    // ".": "./dist/../dist/main.js",       // Invalid: contains ".." segment
    // "./utils/./helper.js": "./utils/helper.js" // Key has invalid segment
  }
}
```

> Node.js resolver (`lib/internal/modules/esm/resolve.js`) checks these.

Reviewer 2: AFAIK, we do not link to node's source-code from the docs.

Author: I have removed it!
